tianshou/policy/modelfree/bdq.py

```diff
@@ -0,0 +1,164 @@
+        indice: np.ndarray,
+        gamma: float = 0.99,
+    ) -> Batch:
+        rew = batch.rew
+        with torch.no_grad():
+            target_q_torch = self._target_q(buffer, indice)  # (bsz, ?)
+        target_q = to_numpy(target_q_torch)
+        end_flag = buffer.done.copy()
+        end_flag[buffer.unfinished_index()] = True
+        end_flag = end_flag[indice]
+        mean_target_q = np.mean(target_q, -1) if len(target_q.shape) > 1 else target_q
+        _target_q = rew + gamma * mean_target_q * (1 - end_flag)
+        target_q = np.repeat(_target_q[..., None], self.num_branches, axis=-1)
+        target_q = np.repeat(target_q[..., None], self.max_action_num, axis=-1)
+
+        batch.returns = to_torch_as(target_q, target_q_torch)
+        if hasattr(batch, "weight"):  # prio buffer update
+            batch.weight = to_torch_as(batch.weight, target_q_torch)
+        return batch
+
+    def process_fn(
+        self, batch: Batch, buffer: ReplayBuffer, indices: np.ndarray
+    ) -> Batch:
+        """Compute the 1-step return for BDQ targets."""
+        batch = self._compute_return(batch, buffer, indices)
+        return batch
+
+    def forward(
+        self,
+        batch: Batch,
+        state: Optional[Union[Dict, Batch, np.ndarray]] = None,
+        model: str = "model",
+        input: str = "obs",
+        **kwargs: Any,
+    ) -> Batch:
+        """Compute action over the given batch data.
+
+        :param float eps: in [0, 1], for epsilon-greedy exploration method.
+
+        :return: A :class:`~tianshou.data.Batch` which has 3 keys:
+
+            * ``act`` the action.
+            * ``logits`` the network's raw output.
+            * ``state`` the hidden state.
+
+        .. seealso::
+
+            Please refer to :meth:`~tianshou.policy.BasePolicy.forward` for
+            more detailed explanation.
+        """
+        model = getattr(self, model)
+        obs = batch[input]
+        obs_next = obs.obs if hasattr(obs, "obs") else obs
+        logits, hidden = model(obs_next, state=state, info=batch.info)
+        act = to_numpy(logits.max(dim=-1)[1])
+        return Batch(logits=logits, act=act, state=hidden)
+
+    def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
+        if self._target and self._iter % self._freq == 0:
+            self.sync_weight()
+        self.optim.zero_grad()
+        weight = batch.pop("weight", 1.0)
+        act = torch.tensor(batch.act)
+        if not batch.returns.get_device() < 0:  # TODO better?
+            act = act.to(batch.returns.get_device())
+        q = self(batch).logits
+        act_mask = torch.zeros_like(q)
+        act_mask = act_mask.scatter_(-1, act.unsqueeze(-1), 1)
+        act_q = q * act_mask
+        returns = batch.returns
+        returns = returns * act_mask
+        td_error = returns - act_q
+        loss = (td_error.pow(2).sum(-1).mean(-1) * weight).mean()
+        # batch.weight = td_error.sum(-1).sum(-1)  # prio-buffer
```

Comment:
Why comment?

Reply:
It's the same as DQN only since the BDQ outputs (bsz, action_branches, action_per_branch) the sum should be done along the last 2 axes. It is commented only because I didn't tested it yet.